ord_schema/validations_test.py

```diff
@@ -0,0 +1,86 @@
+"""Tests for ord_schema.units."""
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+from ord_schema import validations
+from ord_schema.proto import ord_schema_pb2 as schema
+
+
+class UnitsTest(parameterized.TestCase, absltest.TestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.validations = validations
+
+    @parameterized.named_parameters(
+        ('volume', schema.Volume(value=15.0, units=schema.Volume.MILLILITER)),
+        ('time', schema.Time(value=24, units=schema.Time.HOUR)),
+        ('mass', schema.Mass(value=32.1, units=schema.Mass.GRAM)),
+    )
+    def test_units(self, message):
+        self.assertEqual(self.validations.ValidateMessage(message), message)
+
+    @parameterized.named_parameters(
+        ('neg volume', 
+            schema.Volume(value=-15.0, units=schema.Volume.MILLILITER),
+            'non-negative'),
+        ('neg time', schema.Time(value=-24, units=schema.Time.HOUR), 
+            'non-negative'),
+        ('neg mass', schema.Mass(value=-32.1, units=schema.Mass.GRAM),
+            'non-negative'),
+        ('no units', schema.FlowRate(value=5), 'units'),
+        ('percentage out of range', schema.Percentage(value=200), 'between'),
+        ('low temperature', schema.Temperature(value=-5, units='KELVIN'),
+            'between'),
+        ('low temperature 2', schema.Temperature(value=-500, units='CELSIUS'),
+            'between'),
+    )
+    def test_units_should_fail(self, message, expected_error):
+        with self.assertRaisesRegex((ValueError), expected_error):
+            self.validations.ValidateMessage(message)
+
+    def test_orcid(self):
+        message = schema.Person(orcid='0000-0001-2345-678X')
+        self.assertEqual(self.validations.ValidateMessage(message), message)
+
+    def test_orcid_should_fail(self):
+        message = schema.Person(orcid='abcd-0001-2345-678X')
+        with self.assertRaisesRegex((ValueError), 'Invalid'):
+            self.validations.ValidateMessage(message)
+
+    def test_reaction(self):
+        message = schema.Reaction()
+        with self.assertRaisesRegex((ValueError), 'reaction input'):
+            self.validations.ValidateReaction(message)
+
+    def test_reaction_recursive(self):
+        message = schema.Reaction()
+        with self.assertRaisesRegex((ValueError), 'reaction input'):
+            self.validations.ValidateMessage(message, recurse=False)
```

Comment:
In what situations would we want `recurse=False`? I think you'd get the same result here if `recurse` was True?

Reply:
In that case you would get the same error yes. I revised this case to show the behavior I actually wanted to test:

```
message = schema.Reaction()
with self.assertRaisesRegex((ValueError), 'reaction input'):
    validations.ValidateMessage(message)
dummy_input = message.inputs['dummy_input']
self.assertEqual(validations.ValidateMessage(message, recurse=False), 
    message)
```

Once one input is defined, the validation (with ```recurse=False```) is happy and doesn't mind that the input has no components.

I don't know in what situation we would actually choose to use ```recurse=False```, but I wanted to have that flexibility in case it is useful to do on-the-fly validation for the web form just at that level.